ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3PutHandler.java

```diff
@@ -46,19 +48,36 @@ public S3PutHandler(NamedBlobPutHandler namedBlobPutHandler) {
    * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
    */
   public void handle(RestRequest restRequest, RestResponseChannel restResponseChannel, Callback<Void> callback) {
+
+    // 1. Add headers needed that are needed by Ambry
     try {
-      // 1. Add headers needed that are needed by Ambry
       NamedBlobPath namedBlobPath = NamedBlobPath.parse(getRequestPath(restRequest), restRequest.getArgs());
       String accountName = namedBlobPath.getAccountName();
       // Use Account-name as service ID.
       restRequest.setArg(Headers.SERVICE_ID, accountName);
-      // Set x-ambry-content-type to content-type http header
+      // Set x-ambry-content-type from content-type http header
       restRequest.setArg(Headers.AMBRY_CONTENT_TYPE, restRequest.getArgs().get(Headers.CONTENT_TYPE));
-
-      // 2. Upload the blob by following named blob put path
-      namedBlobPutHandler.handle(restRequest, restResponseChannel, callback);
+      // Set x-ambry-content-encoding from content-encoding http header
+      restRequest.setArg(Headers.AMBRY_CONTENT_ENCODING, restRequest.getArgs().get(Headers.CONTENT_ENCODING));
     } catch (RestServiceException e) {
       callback.onCompletion(null, e);
     }
+
+    // 2. Upload the blob by following named blob PUT path
+    namedBlobPutHandler.handle(restRequest, restResponseChannel, (result, exception) -> {
+      try {
```

Comment:
Shall we check the exception and status before we set it to 200?

Reply:
Sure. Updated it to return exception immediately.